internal/wrphandlers/qos/priority_queue.go

```diff
@@ -61,19 +67,33 @@ func (pq *priorityQueue) Enqueue(msg wrp.Message) error {
 	return nil
 }
 
+// trim removes messages with the lowest QualityOfService (taking `prioritizeOldest` into account)
+// until the queue no longer violates `maxQueueSize“.
 func (pq *priorityQueue) trim() {
```

Comment:
I think this is the simplest way we can achieve this while minimizing code changes.

I think the cost of optimized trimming will be small overall because I think that situation where we are trimming often is a very unlikely edge case, it’ll require the xmidt-agent to
-     consistently or completely be unable to send messages to xmidt (e.g outage or connectivity issues)
-     while also receiving a steady flow of new messages to a already full queued

leading to a lot of queue trimming.I think it’s unlikely

But I do believe we can achieve n*log(n) trimming. For example we can manage two queues (each sharing the same pointer to a message) 1 prioritizing the highest QOS and the other  prioritizing the lowest QOS, while `priorityQueue.Dequeue()` handle the logic figuring out if a pop() item is empty (trimmed) or not.

I can throw this idea in a later pr further optimizing if we want to see it.

Reply:
but if we want to optimize trimming further in the future, I think we can achieve n*log(n) trimming (I have a working version locally)